src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs

```diff
@@ -1522,6 +1535,7 @@ private ConsumeNumberResult ConsumeSignMultiSegment(ref ReadOnlySpan<byte> data,
                     }
                     return ConsumeNumberResult.NeedMoreData;
                 }
+                _totalConsumed += i;
```

Comment:
The responsibility of updating this value seems to be in weird places.  Why isn't it updated along with `consumed` in the `return true` path of TryGetNumberMultiSegment instead of in each of the helper pieces?

Reply:
Probably. I wanted to minimize a refactor like this in this PR (particularly because we want this fix -in 3.0). I think some of this logic can be refactored to be clearer (which includes what you mentioned). I had tried to keep the structure of single-segment and multi-segment code similar (and the code has several exit points due to re-entrancy and partial data), which is partially why we have this as a side effect. `consumed` and `_totalConsumed` are coupled due to `GetNextSpan` so I don't want to modify that flippantly. Once we have some more interesting tests covering the multi-segment code paths, I'll feel more confident about a refactor/cleanup like what you are proposing.